src/classes/RD2_OpportunityEvaluationService.cls

```diff
@@ -36,59 +36,134 @@
+    /**
+     * @description Indicates if Opp evaluation service will be skipped.
+     * The flag is currently applicable to migration since Opp evaluation service should not be executed on migration.
+     * For example, if an active RD has no Opp, no new Opp should be created during migration.
+     * Moreover, if the RD is closed, close action on Open Opps should not be applied.
+     */
+    public static Boolean isOverride = false;
 
     /**
-     * @description Recurring Donation Id's to be processed
+    * @description Indicates if the Opportunity evaluation is invoked from a nightly batch job
+    */   
+    private Boolean isBatchContext = false;
+
+    /***
+    * @description Recurring Donation custom settings
+    */
+    private RD2_Settings rdSettings {
+        get {
+            if (rdSettings == null) {
+                rdSettings = new RD2_Settings();
+            } 
+            return rdSettings;
+        }
+        set;
+    }
+
+    /**
+     * @description Indicates if the first installment Opportunity on a Recurring Donation should be created.
+     * If Opportunities are created from a batch, "Disable First Installment" is ignored and 
+     * the first installment is not created only if the setting is "Disable All Installments".
+     * @return Boolean
      */
-    public Set<Id> rdIds {
+    private Boolean isFirstInstallmentEnabled {
         get {
-            if (rdIds == null && rds != null) {
-                rdIds = (new Map<Id, npe03__Recurring_Donation__c>(rds)).keySet();
+            if (isFirstInstallmentEnabled == null) {
+                if (isBatchContext) {
+                    isFirstInstallmentEnabled = !rdSettings.isDisableAllInstallments;
+                } else {
+                    isFirstInstallmentEnabled = rdSettings.isFirstInstallmentEnabled;
+                }
             }
-            return rdIds;
-        } set;
+            return isFirstInstallmentEnabled;
+        }
+        set;
     }
 
     /**
-     * @description Queried Recurring Donation records with subqueries to be processed.
-     */
-    @TestVisible
-    private List<npe03__Recurring_Donation__c> rds {
+    * @description Current reference date to calculate schedules and compare Opps with matching installments
+    */   
+    private Date currentDate {
         get {
-            if (rds == null && rdIds != null) {
-                rds = getRecurringDonationRecords(this.rdIds);
+            if (currentDate == null) {
+                currentDate = RD2_ScheduleService.currentDate;
             }
-            return rds;
+            return currentDate;
         } set;
     }
 
-    @TestVisible
+    /***
+    * @description Recurring Donation Opportunity service
+    */
     private RD2_OpportunityService oppService {
         get {
             if (oppService == null) {
-                oppService = new RD2_OpportunityService();
+                oppService = new RD2_OpportunityService(currentDate, dbService);
```

Comment:
The Opportunity Evaluation service passes the date to Opportunity service. When the eval service is invoked from a batch, then `currentDate` is `Date.today()` by default if not provided with specific date (which should not be outside of testing).
When the eval service is executed from a queuable (after the trigger invoked it), the `currentDate` will be [`RD2_ScheduleService.currentDate`](https://github.com/SalesforceFoundation/NPSP/blob/e199ec537fdab6bcaeb7649d2f19ba27a82af51f/src/classes/RD2_OpportunityEvaluationService.cls#L91) static variable that should correspond to `Date.today()` by default.
So based on this, there is no scenario when Opportunity service date will not be provided. I can change the date to default to `RD2_ScheduelService.currentDate` if null, but not sure if it should be set when eval service date is set since it will be passed in as the argument. Let me know if you think differently.

Reply:
Makes sense. What I really want to be sure of is that a future developer will know why the value is required, what the date value passed into that service will do and what value to pass in.

Reply:
The `current date` descriptions have been added. Thanks @force2b .

Reply:
Thank you!!